lib/puppet/type/hocon_setting.rb

```diff
@@ -22,8 +22,36 @@
     end
   end
 
-  newproperty(:value) do
+  newproperty(:type) do
+    desc "The value type"
+    # This property has no default. If it is not supplied, the validation of the "value"
+    # property will set one automatically.
+  end
+
+  newproperty(:value, :array_matching => :all) do
     desc 'The value of the setting to be defined.'
+
+    validate do |val|
+      case @resource[:type]
+        when nil
+          if @shouldorig.is_a?(Array) and @shouldorig.size > 1
+            @resource[:type] = 'array'
+          else
+            case val.class.to_s.downcase.to_sym
+              when :trueclass, :falseclass
+                @resource[:type] = 'boolean'
+              else
+                @resource[:type] = 'string'
+            end
+          end
+        when 'hash', 'array', 'text'
+          # we're just leaving these values alone
+        else
+          if @shouldorig.is_a?(Array) and @shouldorig.size > 1
+            raise "Array provided, but type specified as #{@resource[:type]}"
+          end
+      end
```

Comment:
Yeah, I agree it's somewhat confusing.

This method handles setting the `type` parameter, but in the provider code, the only specific cases that matter are when it is set to `'array'` or `'text'`. I will probably remove the logic that actually sets that parameter here since it's really not necessary, and just do some basic type validation here with the setting if it happens to be present, since that's what I think should happen anyway.